### ISSUE TITLE
Latest version rnnlm clashes with elephant

### ISSUE THREAD

**Issue by erwanm** [erwanm/elephant-wrapper#29]
Trying to upgrade dependencies as git sub-modules. Apparently something changed with rnnlm (probably related to format) which causes elephant-train to fail when passing the features to wapiti.

**Comment by erwanm** [comment]
Turns out that the options needed to print activation of the hidden layer neurons have simply been removed in the latest version of rnnlm v 0.3d. Visible simply by running elman with no argument, the following were in v 2013.03.27 but are not anymore in 0.3d:

~~~~
        -print-hidden
		Output activations of the hidden layer for the test data
	-hidden-feature-size <int>
		Output only top <int> most active units converted to binary indicator features
~~~~

In other words, rnnlm does not offer backward compatibility ;)

The solution I see is to find the latest version which contains these options and find how to make the git submodule stick to this version.


## version provided with Elephant (v 2013.03.27):

~~~~
Recurrent neural network based language modeling toolkit v 2013.03.27

Options:
Parameters for training phase:
	-train <file>
		Use text data from <file> to train rnnlm model
	-class <int>
		Will use specified amount of classes to decompose vocabulary; default is 100
	-old-classes
		This will use old algorithm to compute classes, which results in slower models but can be a bit more precise
	-rnnlm <file>
		Use <file> to store rnnlm model
	-binary
		Rnnlm model will be saved in binary format (default is plain text)
	-valid <file>
		Use <file> as validation data
	-alpha <float>
		Set starting learning rate; default is 0.1
	-beta <float>
		Set L2 regularization parameter; default is 1e-7
	-hidden <int>
		Set size of hidden layer; default is 30
	-compression <int>
		Set size of compression layer; default is 0 (not used)
	-direct <int>
		Sets size of the hash for direct connections with n-gram features in millions; default is 0
	-direct-order <int>
		Sets the n-gram order for direct connections (max 20); default is 3
	-bptt <int>
		Set amount of steps to propagate error back in time; default is 0 (equal to simple RNN)
	-bptt-block <int>
		Specifies amount of time steps after which the error is backpropagated through time in block mode (default 10, update at each time step = 1)
	-one-iter
		Will cause training to perform exactly one iteration over training data (useful for adapting final models on different data etc.)
	-anti-kasparek <int>
		Model will be saved during training after processing specified amount of words
	-min-improvement <float>
		Set minimal relative entropy improvement for training convergence; default is 1.003
	-gradient-cutoff <float>
		Set maximal absolute gradient value (to improve training stability, use lower values; default is 15, to turn off use 0)
Parameters for testing phase:
	-rnnlm <file>
		Read rnnlm model from <file>
	-test <file>
		Use <file> as test data
	-print-hidden
		Output activations of the hidden layer for the test data
	-hidden-feature-size <int>
		Output only top <int> most active units converted to binary indicator features
	-lm-prob
		Use other LM probabilities for linear interpolation with rnnlm model; see examples at the rnnlm webpage
	-lambda <float>
		Set parameter for linear interpolation of rnnlm and other lm; default weight of rnnlm is 0.75
	-dynamic <float>
		Set learning rate for dynamic model updates during testing phase; default is 0 (static model)
Additional parameters:
	-threads <int>
		Use <int> processors
	-gen <int>
		Generate specified amount of words given distribution from current model
	-independent
		Will erase history at end of each sentence (if used for training, this switch should be used also for testing & rescoring)

Examples:
elman -train train -rnnlm model -valid valid -hidden 50
elman -rnnlm model -test test
~~~~

## Latest version 0.3d:

~~~~
Recurrent neural network based language modeling toolkit v 0.3d

Options:
Parameters for training phase:
	-train <file>
		Use text data from <file> to train rnnlm model
	-class <int>
		Will use specified amount of classes to decompose vocabulary; default is 100
	-old-classes
		This will use old algorithm to compute classes, which results in slower models but can be a bit more precise
	-rnnlm <file>
		Use <file> to store rnnlm model
	-binary
		Rnnlm model will be saved in binary format (default is plain text)
	-valid <file>
		Use <file> as validation data
	-alpha <float>
		Set starting learning rate; default is 0.1
	-beta <float>
		Set L2 regularization parameter; default is 1e-7
	-hidden <int>
		Set size of hidden layer; default is 30
	-compression <int>
		Set size of compression layer; default is 0 (not used)
	-direct <int>
		Sets size of the hash for direct connections with n-gram features in millions; default is 0
	-direct-order <int>
		Sets the n-gram order for direct connections (max 20); default is 3
	-bptt <int>
		Set amount of steps to propagate error back in time; default is 0 (equal to simple RNN)
	-bptt-block <int>
		Specifies amount of time steps after which the error is backpropagated through time in block mode (default 10, update at each time step = 1)
	-one-iter
		Will cause training to perform exactly one iteration over training data (useful for adapting final models on different data etc.)
	-max-iter
		Will cause training to perform exactly <max-iter> iterations over training data (useful to test static learning rates if min-improvement is set to 0.0)
	-anti-kasparek <int>
		Model will be saved during training after processing specified amount of words
	-min-improvement <float>
		Set minimal relative entropy improvement for training convergence; default is 1.003
	-gradient-cutoff <float>
		Set maximal absolute gradient value (to improve training stability, use lower values; default is 15, to turn off use 0)
Parameters for testing phase:
	-rnnlm <file>
		Read rnnlm model from <file>
	-test <file>
		Use <file> as test data to report perplexity
	-lm-prob
		Use other LM probabilities for linear interpolation with rnnlm model; see examples at the rnnlm webpage
	-lambda <float>
		Set parameter for linear interpolation of rnnlm and other lm; default weight of rnnlm is 0.75
	-dynamic <float>
		Set learning rate for dynamic model updates during testing phase; default is 0 (static model)
Additional parameters:
	-gen <int>
		Generate specified amount of words given distribution from current model
	-independent
		Will erase history at end of each sentence (if used for training, this switch should be used also for testing & rescoring)

Examples:
rnnlm -train train -rnnlm model -valid valid -hidden 50
rnnlm -rnnlm model -test test
~~~~

**Comment by erwanm** [comment]
Apparently even the first version 0.1h listed on the [original repo](https://github.com/mspandit/rnnlm) (http://www.fit.vutbr.cz/%7Eimikolov/rnnlm/rnnlm-0.1h.tgz) does not include the necessary options. Also looked at a couple other versions, no luck. This means that the version in Elephant is the only one with these options, afaik.

**Comment by erwanm** [comment]
After checking, it looks like the Elephant version is actually a fork by Grzegorz Chrupała:

~~~~
///////////////////////////////////////////////////////////////////////
//
// Recurrent neural network based statistical language modeling toolkit
// Version 2013.04.10
// (c) 2010-2012 Tomas Mikolov (tmikolov@gmail.com)
// (c) Grzegorz Chrupała (g.chrupala@uvt.nl)
///////////////////////////////////////////////////////////////////////
~~~~

**Comment by erwanm** [comment]
Ok, I was actually mistaken about the genealogy tree of this software! It is indeed a fork by Grzegorz Chrupała, the official version is at https://bitbucket.org/gchrupala/elman

**Comment by erwanm** [comment]
done in 0.1
